lib/Sema/CSDiag.cpp

```diff
@@ -5599,6 +5599,60 @@ static bool shouldTypeCheckFunctionExpr(TypeChecker &TC, DeclContext *DC,
   return true;
 }
 
+// Check if any of the candidates can accept a specified number of arguments,
+// regardless of parameter type or label information.
+static bool acceptNumArgs(const CalleeCandidateInfo &CCI, size_t numArgs) {
```

Comment:
I think it would be great if logic that we already have in CSRanking could be extracted and re-used here as well, instead - https://github.com/apple/swift/blob/master/lib/Sema/CSRanking.cpp#L598-L701 it also handles trailing closures.

Reply:
Thanks for leading the way to the relevant code, Pavel! It is not straight forward to reuse, though.

The logic seems to be different. Let's say we have a candidate with variadic parameter `foo(_: Int...)`, and it accepts the argument list `(1, 2, 3)`. The `acceptNumArgs` function here returns true, while the code snippet in `CSRanking.cpp` would not due to [this line](https://github.com/apple/swift/blob/master/lib/Sema/CSRanking.cpp#L610).

In addition, the code snippet is a bit tightly coupled, with some variables defined in outer scope, and some control flow statements, like [here](https://github.com/apple/swift/blob/master/lib/Sema/CSRanking.cpp#L671) and [here](https://github.com/apple/swift/blob/master/lib/Sema/CSRanking.cpp#L686)).

The problem is that do we really have to handle trailing closure? If the the argument number mismatch happens in the closure, this PR does not handle it and compiler's behaviour remains unchanged. I mean, it does not make things worse.

Reply:
Sorry @dingobye I didn't mean that it has to be quiet literally the same, but I think it makes sense to extract common logic and leave irrelevant checks in their original places. I think the logic could be - given array of parameters can it match given array of arguments, maybe with additional callback to process each accepted argument/parameter pair further, this way it could be used by filter* functions in CalleeCandidateInfo as well.

Reply:
I am trying to extract common logic, but have some troubles. It seems to produce an opposite result for the following case:

Let's say we have `params1 = (_: Int, _: Int)` and `params2 = (_: Int...)`. The code in CSRank.cpp concludes `params1` is NOT a subtype of `params2`. Please check [this line](https://github.com/apple/swift/blob/master/lib/Sema/CSRanking.cpp#L610). However, I would expect the opposite, i.e., a given array of parameters `params2` can match a given array of arguments with `params1`'s type (e.g., `(0, 0)`). Because a variadic Int parameter should match a list of Int arguments.

I get confused here. Did I miss something?

Reply:
@dingobye I think it's correct behavior in this case because (according to comment) it tries to figure out if parameters from first function could be forwarded to the second, which in case of one having 2 parameters and another having a single variadic can't happen, same goes for the arguments with labels as well, they can't be forwarded to a single variadic parameter e.g.

```swift

func foo(_ params: Int...) {} // signature of parameters is (params: Int...)
foo(x: 0, y: 1) // fail - signature of arguments is (x: Int, y: Int)
```

Reply:
I think it should be attempted as hard as possible to avoid code duplications in multiple places in the solver, especially when it has to do with parameter matching. As I mentioned before I think we can have processing of each parameter done via callback but defaults/variadics could be handled in generalized way, also as a more general remark I think there is a better way then trying to compare `CalleeCandidateInfo`s, although it would probably require more changes to archive.

Reply:
Hi Pavel, I updated the code as per your suggestions. Some logics are extracted, with variadic and defaulted parameters handled in general, and specific matching behaviours (e.g., type checking) are done via callback.

Reply:
Thank you, @dingobye! I'll try to take a look over the weekend.

Reply:
Do you think it makes sense to rename this function to something more verbose like `isViableOverloadSet`?

Reply:
Good idea! I was not so happy with the name.